### PR TITLE
Wire orchestrator-mode mint authorization signing and delivery

### PR DESCRIPTION
**Pull request by rouzwelt**
## What

Implements orchestrator-mode mint authorization (RAI-1243). When an asset's `vault_mode` on the issuance service reads `"orchestrator"`, the mint saga now signs a `MintAuthV1` recipient authorization and delivers it to issuance before polling for completion. Vault-direct assets are unaffected.

## Why

Issuance will not execute an orchestrator-mode mint until it receives a valid recipient authorization signed by the configured orchestrator wallet. Without this, orchestrator-mode mints would be accepted by the bot but stall indefinitely at the issuance polling step.

## How

**Aggregate changes (`tokenized_equity_mint.rs`):**

- Added `MintAuthorizationProgress` enum (`NotSigned`, `Signed`, `Delivered`) embedded in the `MintAccepted` state with `#[serde(default)]` so existing event streams replay unchanged.
- Added `MintAuthorizationSigned` and `MintAuthorizationDelivered` events, persisted to the event log with hex-encoded nonce and signature.
- Added `SignMintAuthorization` and `RecordMintAuthorizationDelivered` commands. Signing is idempotent: a re-drive finding an existing authorization emits nothing rather than minting a fresh nonce (the nonce is the mint's on-chain idempotency key).
- Bumped `SCHEMA_VERSION` to 5 to clear stale snapshots.
- Extended `interrupted_mint_ids` to include mints whose latest event is `MintAuthorizationSigned` or `MintAuthorizationDelivered`, so a crash between signing and delivery is recovered on restart.

**`ConfiguredMintAuthorizer`:**

- New `Enabled`/`Disabled` enum added to `mint_authorization.rs`, mirroring `BotGasReceiptCostEnqueuer`. `Disabled` fails loudly if signing is ever reached without an `[orchestrator]` config section. Added to `EquityTransferServices`.

**Saga wiring (`rebalancing/equity/mod.rs`):**

- `CrossVenueEquityTransfer` gains an optional `MintAuthorizationWiring` (vault-mode reader, token address map, delivery queue), opted in via `with_mint_authorization`.
- `ensure_mint_authorization` is called both on the initial mint acceptance path and on every resume from `MintAccepted`. It reads `vault_mode` from issuance, signs if orchestrator-mode, and enqueues the delivery job. A live-delivery guard prevents a second delivery chain from starting during a delayed-redrive window.

**Delivery job (`rebalancing/equity/authorization_job.rs`):**

- New `DeliverMintAuthorization` apalis job. Loads the persisted authorization from the aggregate and delivers it to issuance. Uses bounded delayed redrives (30 s delay, 20-attempt budget, ~10 minutes total) for retryable outcomes (`MintNotFoundYet`, transport failures) rather than consuming the three-attempt apalis budget. Non-retryable rejections (409/422) park the mint with an operator alert and return `Ok` so the best-effort worker keeps running.
- Registered on a best-effort worker so a stuck authorization never halts hedging or fill detection.

**Conductor wiring:**

- `build_mint_authorization_infra` constructs the `ConfiguredMintAuthorizer` (enabled only when `[orchestrator]` is configured), the delivery queue, the `MintAuthorizationWiring`, and a shared `IssuanceClient`. Orphaned delivery rows from a prior crash are requeued at startup.
- `DeliverMintAuthorizationJobQueue` and `DeliverMintAuthorizationCtx` are threaded through `PositionAndRebalancing`, `RebalancingInfrastructure`, and the conductor builder.
- CLI and non-rebalancing paths set `mint_authorizer: ConfiguredMintAuthorizer::Disabled`; the comment explains why orchestrator-mode mints must run through the server.

**Stall clock and performance tracking:**

- `MintAuthorizationSigned` and `MintAuthorizationDelivered` refresh the stall clock in `MintTracking` without advancing the stage (genuine forward progress for a mint waiting in `Accepted`).
- Both events are treated as no-ops in inventory, reliability, and timing projections where they carry no share movement.

## Testing

- Unit tests on the aggregate cover: signing persists `Signed` progress; re-signing keeps the original nonce; distinct mints generate independent nonces; signing with `Disabled` authorizer fails loudly; delivery transitions to `Delivered` and is idempotent; delivery without a signed authorization errors.
- Wire-format pin tests assert the exact hex JSON shape of `MintAuthorizationSigned` and `MintAuthorizationDelivered` events.
- Replay tests confirm pre-authorization streams default to `NotSigned` and that a pinned hex payload materializes with the correct nonce and signature bytes.
- `interrupted_mint_ids` test confirms mints awaiting delivery are included in startup recovery.
- Delivery job tests cover: `Recorded` marks the aggregate delivered; `Rejected`/`Conflict` parks with one operator alert; `MintNotFoundYet` schedules a delayed redrive; budget exhaustion dead-letters with an alert; an already-delivered mint is a no-op.
- Integration tests on `CrossVenueEquityTransfer` cover: orchestrator-mode resume signs once and enqueues exactly one delivery job; a second resume is idempotent; the live-delivery guard prevents a second chain over a pending redrive successor; vault-direct resume signs nothing and enqueues nothing.

## Anything else

The `[orchestrator]` config section remains optional. While every asset is vault-direct the bot deploys without it, and an orchestrator-mode mint reaching the signing step fails loudly rather than guessing an address. The mint stays in `MintAccepted` with its authorization persisted and is resumable once the section is added.

Closes [RAI-1677](https://linear.app/makeitrain/issue/RAI-1677/mint-authorization-signing-delivery-and-mode-discovery)  
Part of [RAI-1243](https://linear.app/makeitrain/issue/RAI-1243/liquidity-bot-sign-and-deliver-mintauthv1-for-orchestrator-mode-mints)

---

<picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture> <picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture>
<sup>Need help on this PR? Tag `@codesmith-bot` with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

* **New Features**
  * Added orchestrator-mode mint authorization with signing, persistence, and delivery acknowledgements.
  * Added stable per-mint nonces, signature reuse, idempotent delivery, and duplicate protection.
  * Added automatic retries for temporary delivery failures and operator alerts for unresolved authorizations.
  * Vault-direct mints continue without authorization delivery.
* **Bug Fixes**
  * Authorization events now update mint progress and timing without advancing lifecycle stages or reporting failures.
  * Improved recovery handling for interrupted authorization deliveries.

### CODE REVIEW

**Comment by rouzwelt** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1145?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1209** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1209?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1202** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1202?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1151** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1151?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1150** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1150?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1145** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1145?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1145?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1141** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1141?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1139** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1139?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1138** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1138?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1130** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1130?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1129** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1129?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1128** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1128?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1115** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1115?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1114** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1114?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1113** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1113?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>add-to-gt-merge-queue</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/ST0x-Technology?ref=merge-queue-instructions-in-stack-comment&prId=1145).

<h2></h2>
<sub>

An organization admin has required the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by rouzwelt** [comment]
@coderabbit review

File: src/conductor.rs [hunk 1]
```diff
@@ -1769,6 +1770,128 @@ fn build_alert_notifier(
...
+        wiring: MintAuthorizationWiring {
+            vault_mode_reader: issuance_client.clone(),
+            token_addresses,
+            delivery_queue: queue.clone(),
+        },
+        queue,
+        issuance_client,
+    })
+}
+
```
[the remark this answers is not in the document]
Comment (rouzwelt) [reply src/conductor.rs:1957]: both branches are already covered end-to-end — the orchestrator e2e configures [orchestrator] and exercises the Enabled path through a real signed delivery, and every other e2e boots without the section through Disabled. A unit test of this function would require constructing a full Ctx to assert DI plumbing (queue clones reaching three fields), which the compiler already guarantees structurally. Not worth the fixture weight.

File: src/conductor.rs [hunk 2]
```diff
@@ -2057,12 +2180,16 @@ fn spawn_rebalancing_infrastructure<Chain: Wallet + Clone>(
...
+        let mint_authorization =
+            build_mint_authorization_infra(&deps.ctx, &deps.apalis_pool, &base_wallet).await?;
+
         let equity_transfer_services = EquityTransferServices {
             raindex: raindex_service.clone(),
             vault_lookup: vault_lookup.clone(),
             tokenizer: tokenizer.clone(),
             wrapper: wrapper.clone(),
             bot_gas_enqueuer: bot_gas_enqueuer.clone(),
+            mint_authorizer: mint_authorization.authorizer,
```
[the remark this answers is not in the document]
Comment (rouzwelt) [reply src/conductor.rs:2257]: The cited learning doesn't match the current codebase: with_bot_gas_enqueuer is itself still a builder method on this exact struct today, so the live precedent is builder-with-explicit-absence-default. This PR now strengthens that shape — absence became an explicit ConfiguredMintAuthorization::VaultDirectOnly variant (mirroring BotGasReceiptCostEnqueuer::Disabled) rather than None. Switching to a constructor parameter would touch ~28 construction sites for consistency with a convention the code doesn't follow; happy to do that as its own sweep covering bot-gas too if we want to adopt it for real.

File: src/tokenized_equity_mint.rs [hunk 3]
```diff
@@ -1577,6 +1780,9 @@ impl EventSourced for TokenizedEquityMint {
                     tokenization_request_id,
                     requested_at,
                     accepted_at,
+                    // The authorization served its purpose once tokens
+                    // arrive; later states deliberately do not carry it.
+                    authorization: _,
```
[the remark this answers is not in the document]
Comment (rouzwelt) [reply src/tokenized_equity_mint.rs:1785]: The analysis is against an older revision. transition_record_authorization_delivered no longer requires a Signed state — a post-advance acknowledgement records unconditionally as an audit fact (see recording_delivery_after_mint_advanced_still_records). And the job's post-advance no-op is correct, not a gap: the mint can only reach TokensReceived after issuance minted, which it does only once the authorization is already in its hands — so there is nothing left to deliver, and dropping the payload with the state is intentional.

File: src/cli/rebalancing.rs [hunk 4]
```diff
@@ -136,6 +137,17 @@ async fn build_equity_transfer_services(
...
+        // so the saga cannot read `vault_mode` and treats every asset as
+        // vault-direct (with a warning) -- a CLI mint of an
+        // orchestrator-mode asset therefore proceeds unauthorized and
+        // stalls at polling until the authorization is delivered by the
+        // server's resume path. This `Disabled` value is only consulted if
+        // signing is ever reached, which the missing wiring prevents.
+        mint_authorizer: ConfiguredMintAuthorizer::Disabled,
     };
 
     let mint_store = StoreBuilder::<TokenizedEquityMint>::new(pool.clone())
```
Comment (0xgleb) [inline src/cli/rebalancing.rs:152]: should fix: This CLI path can initiate an orchestrator-mode mint with no vault-mode reader, signer, or delivery queue. Reject the command before acceptance or wire it through the same authorization infrastructure as the server.

File: src/rebalancing/equity/mod.rs [hunk 5]
```diff
@@ -681,6 +753,117 @@ impl CrossVenueEquityTransfer {
...
+                    )
+                    .await?;
+
+                // Bounds delivery to ONE live chain per mint: redrive
+                // successors carry no idempotency key, so without this
+                // check a resume during a delayed-redrive window would
+                // start a second chain. Fails open on a guard read error
+                // (warn + push): a duplicate chain delivers byte-identical
+                // payloads and is benign, while a skipped push would stall
+                // the delivery.
```
Comment (0xgleb) [inline src/rebalancing/equity/mod.rs:818]: minor: Propagate failures from the active-delivery lookup instead of treating them as `false`. The check and enqueue sequence can still race, so a per-mint uniqueness or idempotency guard would make the intended one-live-chain invariant durable, even though duplicate chains deliver the same payload.

File: src/conductor.rs [hunk 6]
```diff
@@ -1769,6 +1770,128 @@ fn build_alert_notifier(
...
+    /// One issuance client serves both mint-authorization consumers (the
+    /// saga's vault-mode read and the delivery job), built from the same
+    /// `[issuance]` credentials as the freeze guard's own client instance.
+    issuance_client: Arc<IssuanceClient>,
+}
+
+/// Builds [`MintAuthorizationInfra`]. The authorizer is `Enabled` only when
+/// `[orchestrator]` is configured: while every asset is vault-direct the bot
+/// deploys dark without the section, and an orchestrator-mode mint reaching
+/// the signing step then fails loudly rather than guessing an address.
```
Comment (0xgleb) [inline src/conductor.rs:1889]: should fix: Continuing after orphan repair fails is unsafe because a stale `Running` row is then considered live and suppresses resume enqueueing. Fail startup or recover stale rows by lease age before allowing that suppression.

File: src/rebalancing/equity/authorization_job.rs [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+/// alone cannot see them -- without this bound, a restart during a
+/// delayed-redrive window would start a second delivery chain. Chains
+/// deliver byte-identically (issuance-idempotent), so a duplicate is
+/// benign rather than dangerous; this mirrors the one-poll-job-per-order
+/// bound on `PollOrderStatus`. "Live" matches `requeue_orphaned`'s
+/// in-flight definition: `Pending`, `Queued` (claimed, not yet running),
+/// or `Running`.
+pub(crate) async fn has_live_delivery_job(
+    queue: &DeliverMintAuthorizationJobQueue,
+    issuer_request_id: &IssuerRequestId,
```
Comment (0xgleb) [inline src/rebalancing/equity/authorization_job.rs:309]: should fix: `Conflict` and `Rejected` currently leave only logs and an alert, so startup recovery can enqueue the same non-retryable delivery again. Persist a parked outcome before completing the job, exclude it from automatic recovery, and cover both branches.

File: src/rebalancing/equity/authorization_job.rs [hunk 8]
```diff
@@ -0,0 +1,648 @@
...
+
+/// Alert delivery failures are logged, never allowed to mask the job
+/// outcome (mirrors the USDC transfer jobs' alert handling).
+async fn notify_swallowing_failure(ctx: &DeliverMintAuthorizationCtx, message: &str) {
+    if let Err(alert_error) = ctx.notifier.notify(message).await {
+        warn!(
+            target: "tokenization",
+            ?alert_error,
+            "Failed to deliver mint-authorization operator alert"
+        );
```
Comment (0xgleb) [inline src/rebalancing/equity/authorization_job.rs:360]: should fix: Returning this error after exhausting redrives lets the generic worker retry the same exhausted payload and repeat the alert. Persist exhaustion and complete successfully, or make this failure non-retryable with durable alert deduplication.

File: src/tokenized_equity_mint.rs [hunk 9]
```diff
@@ -2137,6 +2351,175 @@ impl EventSourced for TokenizedEquityMint {
...
+            }
+            | Self::WrapSubmitted {
+                issuer_request_id, ..
+            }
+            | Self::TokensWrapped {
+                issuer_request_id, ..
+            }
+            | Self::VaultDepositSubmitted {
+                issuer_request_id, ..
+            }
```
Comment (0xgleb) [inline src/tokenized_equity_mint.rs:2497]: minor: Retain the authorization identity, or include it in `RecordMintAuthorizationDelivered`, so a late acknowledgement can be checked against the mint that was signed. The current later-state arms accept the acknowledgement after authorization progress is dropped, which weakens audit provenance but does not itself alter issuance or custody.

File: src/tokenized_equity_mint.rs [hunk 10]
```diff
@@ -2137,6 +2351,175 @@ impl EventSourced for TokenizedEquityMint {
...
+                let nonce = B256::random();
+                let signed = services
+                    .mint_authorizer
+                    .sign(token, *quantity, nonce)
+                    .await
+                    .map_err(
+                        |error| TokenizedEquityMintError::AuthorizationSigningFailed {
+                            error_message: error.to_string(),
+                        },
+                    )?;
```
Comment (0xgleb) [inline src/tokenized_equity_mint.rs:2397]: minor: Preserve `MintAuthorizationError` as a typed source instead of converting it to a string. This keeps configuration, scaling, and EVM failures distinguishable and preserves the source chain.

File: src/conductor/builder.rs [hunk 11]
```diff
@@ -1195,6 +1214,37 @@ fn register_resume_tokenization_worker(
     })
```
Comment (0xgleb) [inline src/conductor/builder.rs:1240]: minor: Add a builder-level regression test covering a delivery context with a queued authorization job, plus the intentional `None` context case. The registration change is straightforward, so this is coverage hardening rather than evidence that the worker is currently omitted.

**Review by 0xgleb — CHANGES_REQUESTED** [review]

File: src/rebalancing/equity/mod.rs (hunk 5 above)
Comment (rouzwelt) [reply src/rebalancing/equity/mod.rs:818]: Fail-open is deliberate and the asymmetry is the reason: a duplicate chain delivers byte-identical payloads that issuance treats as an idempotent 200, so the worst case of proceeding is benign, while propagating a transient DB read error would fail the transfer step and stall the mint. Durable uniqueness already exists where it matters — the initial enqueue is push_idempotent keyed on the issuer request id, so the check-then-enqueue race can only ever duplicate a redrive successor, which is the benign case above.

File: src/rebalancing/equity/authorization_job.rs (hunk 7 above)
Comment (rouzwelt) [reply src/rebalancing/equity/authorization_job.rs:309]: Excluding parked mints from automatic recovery would break the resolution path: resolution happens on the issuance side, and the resume-driven flow is how this bot observes it (polling into TokensReceived). Redelivery after a park is byte-identical and issuance-idempotent — a still-unresolved 409 just re-parks with one alert per restart, which is deliberate: an unresolved park should resurface rather than go quiet. Persisting a parked flag in the aggregate buys alert dedup across restarts at the cost of an event-schema addition plus a mechanism to clear it once issuance resolves — happy to file it as a follow-up if restart-frequency alert noise shows up in practice.

File: src/tokenized_equity_mint.rs (hunk 9 above)
Comment (rouzwelt) [reply src/tokenized_equity_mint.rs:2497]: Provenance lives in the event stream, not the state: MintAuthorizationSigned (nonce + signature) is permanently in the same per-mint stream the late MintAuthorizationDelivered lands in, so the correlation is the stream identity itself. There's also no cross-mint path to guard against — the ack is dispatched by the delivery job keyed on this aggregate's id, delivering the authorization it just loaded from this aggregate. Adding the nonce to the ack event would record a value the handler has nothing to check it against once state has advanced (the authorization is dropped precisely because issuance holds it).

File: src/tokenized_equity_mint.rs (hunk 10 above)
Comment (rouzwelt) [reply src/tokenized_equity_mint.rs:2397]: Agreed in principle — this is the one place the repo's no-string-errors rule bends, and it's a framework constraint, not a choice: TokenizedEquityMintError must satisfy the DomainError serde bound, and MintAuthorizationError wraps EvmError (non-serde transport internals), so the typed source can't be embedded. Same documented constraint as the neighboring RequestFailed (see the variant's doc comment). The rendered message preserves the full Display chain (config/scaling/EVM failures stay distinguishable in logs), and no caller branches on the variant.

File: src/conductor/builder.rs (hunk 11 above)
Comment (rouzwelt) [reply src/conductor/builder.rs:1240]: None of the six conditionally-registered sibling workers has a registration-level test, for a structural reason: apalis Monitor doesn't expose its registrations for assertion, so such a test could only re-assert the guard clause's own code. The Some path is exercised end-to-end in the stacked e2e PR (the delivery worker actually consumes the queue and the mint completes), and the None arm is a debug-log guard identical to its five siblings. Happy to add it if we introduce a registration-introspection seam, but that's a builder-wide change, not specific to this worker.

**Review by 0xgleb — APPROVED** [review]

**Review by JuaniRios — APPROVED** [review]
